crates/blockchain/blockchain.rs

```diff
@@ -55,109 +55,131 @@ impl Blockchain {
-        let since = Instant::now();
-
-        let block_hash = block.header.compute_block_hash();
+    pub fn add_block(&self, block: Block) -> Result<(), ChainError> {
+        self.add_blocks_in_batch(vec![block], false)
+    }
 
-        // Validate if it can be the new head and find the parent
-        let Ok(parent_header) = find_parent_header(&block.header, &self.storage) else {
-            // If the parent is not present, we store it as pending.
-            self.storage.add_pending_block(block.clone())?;
-            return Err(ChainError::ParentNotFound);
+    pub fn add_blocks_in_batch(
+        &self,
+        blocks: Vec<Block>,
+        should_commit_intermediate_tries: bool,
+    ) -> Result<(), ChainError> {
+        let first_block_header = match blocks.first() {
+            Some(block) => block.header.clone(),
+            None => return Err(ChainError::Custom("First block not found".into())),
+        };
+        let last_block_header = match blocks.last() {
+            Some(block) => block.header.clone(),
+            None => return Err(ChainError::Custom("Last block not found".into())),
         };
 
-        let chain_config = self.storage.get_chain_config()?;
+        if self.evm_engine == EvmEngine::LEVM {
+            panic!("LEVM engine is not supported for add blocks in batch");
+        }
 
-        // Validate the block pre-execution
-        validate_block(block, &parent_header, &chain_config)?;
+        let Some(mut state_trie) = self.storage.state_trie(first_block_header.parent_hash)? else {
+            return Err(ChainError::ParentNotFound);
+        };
 
+        let mut all_receipts: Vec<(BlockHash, Vec<Receipt>)> = vec![];
+        let chain_config: ChainConfig = self.storage.get_chain_config()?;
         let mut vm = Evm::new(
             self.evm_engine,
             self.storage.clone(),
-            block.header.parent_hash,
+            first_block_header.parent_hash,
         );
-        let BlockExecutionResult {
-            receipts,
-            requests,
-            account_updates,
-        } = vm.execute_block(block)?;
 
-        validate_gas_used(&receipts, &block.header)?;
+        for (i, block) in blocks.iter().enumerate() {
+            let is_first_block = i == 0;
+            let is_last_block = i == blocks.len() - 1;
+
+            let block_hash = block.header.compute_block_hash();
+
+            // Validate if it can be the new head and find the parent
+            let parent_header = match is_first_block {
+                // for the first block, we need to query the store
+                true => {
+                    let Ok(parent_header) = find_parent_header(&block.header, &self.storage) else {
+                        // If the parent is not present, we store it as pending.
+                        self.storage.add_pending_block(block.clone())?;
+                        return Err(ChainError::ParentNotFound);
```

Comment:
Shouldn't the remaining blocks be stored as pending too? We are losing track of them here